source/MaterialXTest/ShaderValid.cpp

```diff
@@ -1022,8 +1060,118 @@ TEST_CASE("MaterialX documents", "[shadervalid]")
+#endif
+        size_t skipCount = 0;
+        profilingLog << "-- Possibly missed implementations ----" << std::endl;
+        std::vector<std::string> whiteList =
+        {
+            "arrayappend", "backfacing", "screen", "curveadjust", "dot_surfaceshader", "mix_surfaceshader"
+            "displacementShader", "displacementshader", "volumeshader", "IM_dot_filename", "ambientocclusion", "dot_lightshader"
+        };
+        const std::string OSL_STRING("osl");
+        const std::string SX_OSL_STRING("sx_osl");
+        unsigned int implementationUseCount = 0;
+        for (auto libraryImpl : libraryImpls)
+        {
+            const std::string& implName = libraryImpl->getName();
+
+            // Skip white-list items
+            bool whileListFound = false;
+            for (auto w : whiteList)
+            {
+                if (implName.find(w) != std::string::npos)
+                {
+                    skipCount++;
+                    whileListFound = true;
+                    break;
+                }
+            }
+            if (whileListFound)
+            {
+                implementationUseCount++;
+                continue;
+            }
+
+            if (usedImpls.count(implName))
+            {
+                implementationUseCount++;
+                continue;
+            }
+
+            if (oslShaderGenerator && oslShaderGenerator->getCachedImplementation(implName))
+            {
+                implementationUseCount++;
+                continue;
+            }
+            if (glslShaderGenerator && glslShaderGenerator->getCachedImplementation(implName))
+            {
+                implementationUseCount++;
+                continue;
+            }
+            if (oslCms && oslCms->getCachedImplementation(implName))
+            {
+                implementationUseCount++;
+                continue;
+            }
+            if (glslCms && glslCms->getCachedImplementation(implName))
+            {
+                implementationUseCount++;
+                continue;
+            }
+
+            // See if we have a sx-osl implementation used
+            // instead of the reference one
+            if (libraryImpl->getLanguage() == OSL_STRING)
+            {
+                size_t endSize = implName.size() - 3;
+                if (endSize > 0)
+                {
+                    std::string ending = implName.substr(endSize);
+                    if (ending == OSL_STRING)
+                    {
+                        std::string sxImplName = implName.substr(0, endSize) + SX_OSL_STRING;
+                        if (oslShaderGenerator->getCachedImplementation(sxImplName))
+                        {
+                            implementationUseCount++;
+                            continue;
+                        }
+                    }
+                }
+            }
+
+            profilingLog << "\t" << implName << std::endl;
```

Comment:
This will dump out missing ones.

Reply:
Ah, I missed this log output.

Reply:
Would this work when a language is disabled? For example if OSL is disabled by build variables we would not have an osl generator and the check above would not catch an OSL implementation element loaded in the library. I think all languages are always loaded even if the shader generation for it is disabled.
I guess same question for CMS, if no CMS is set on the generator it would not catch the CMS implementation elements loaded in the library.

Reply:
No this will show more missing ones if a language was disabled. stdlib always loads so it will have the reference impls, so even if the GLSL impls and OSL impls could be selectively loaded it will be hard to do the matching. I was going to make GLSL and OSL always run if you wanted to do the implementation check. I will make this change.

CMS should only load their libraries in conjunction with setting them on the generator, so that will need to be the current assumption. I guess the safer way would be do this on demand but I think I'll leave this for Jonathan to add when he adds in actual tests as it would be best to load the library the first time a color transform is required which would be detected within the shadergen code.

Reply:
OK. Yeah I agree CMS should only load on demand, I don't think that's the case currently.

I'll approve this review now so I'm not blocking you :) We can always refine this later if needed.